File: ruby/src/data_structures/arrays/sum_pair.rb
#Design an algorithm to find all pairs of integers within an array which sum to a specified value.

def find_pairs(arr, sum)
  hash = {}
  arr.each do |a|
    hash[a] = 0 unless hash.key?(a)
    hash[a] += 1
  end
  arr.each do |a|
    value = hash[sum - a]
    next unless value
    next if a == (sum-a) && value < 2
    puts a, sum-a
    hash[a] -= 1
    hash[sum-a] -= 1
    hash.delete(a) if hash[a] <= 0
    hash.delete(sum-a) if hash[sum-a] && hash[sum-a] <= 0
  end
end

def find_pair_sorted_array(arr, sum)
  arr.sort!
  first = 0
  last = arr.size-1
  while (first < last)
    current_sum = arr[first] + arr[last]
    if current_sum == sum
      print arr[first], " ", arr[last], "\n"
      first += 1
      last -= 1
    elsif current_sum < sum
      first += 1
    else
      last -= 1
    end
  end
end

puts "find 1"
find_pairs([4,4,4,4], 8)
puts "find 2"
find_pairs([4,5], 8)
puts "find 3"
find_pairs([-3,4,5,1,2,3,-5], 0)
puts "find 1"
find_pair_sorted_array([4,4,4,4], 8)
puts "find 2"
find_pair_sorted_array([4,5], 8)
puts "find 3"
find_pair_sorted_array([-3,4,5,1,2,3,-5], 0)

File: java/src/data_structures/arrays/SumPair.java
package src.data_structures.arrays;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

//Design an algorithm to find all pairs of integers within an array which sum to a specified value.
public class SumPair {
	public List<Integer[]> findAllPairs(int[] arr, int sum) {
		List<Integer[]> result = new ArrayList<Integer[]>();
		Map<Integer, Integer> map = new HashMap<Integer, Integer>();
		
		for (int element: arr) {
			if (map.containsKey(sum - element)) {
				// Found a pair
				Integer[] pair = {element, sum-element};
				result.add(pair);
				int currentCount = map.get(sum - element);
				currentCount--;
				if (currentCount <= 0) {
					map.remove(sum - element);
				} else {
					map.put(sum - element, currentCount); // Storing for multiple occurrence of same numbers
				}
			} else {
				map.put(element, 1);
			}
		}
		
		return result;
	}
	
	public List<Integer[]> findAllPairsNew(int[] arr, int sum) {
		Arrays.sort(arr);
		return findAllPairsSort(arr, sum);
	}
	
	public List<Integer[]> findAllPairsSort(int[] arr, int sum) {
		List<Integer[]> result = new ArrayList<Integer[]>();
		
		for (int i=0,j=arr.length-1; i<j; ) {
			if (arr[i] + arr[j] == sum) {
				Integer[] pair = {arr[i], arr[j]};
				result.add(pair);
				i++;
				j--;
			} else if (arr[i] + arr[j] > sum) {
				j--;
			} else {
				i++;
			}
		}
		
		return result;
	}
	
	
	public static void main(String[] args) {
		int[] arr = {2,3,11,8,0,9,-1,3,12, 8};
		List<Integer[]>result = new SumPair().findAllPairsNew(arr, 11);
		for (Integer[] pair: result) {
			System.out.println(pair[0] + ", " + pair[1]);
		}
	}
}
